### PR TITLE
Expose and add unit tests for ISOWeek

### LINKED ISSUE #28933: Add GetISOWeekOfYear()
## Rationale

.Net don't implement [ISO 8601 standard](https://en.wikipedia.org/wiki/ISO_week_date#Relation_with_the_Gregorian_calendar) - it have alternative scheme in:
```c#
System.Globalization.Calendar.GetWeekOfYear(DateTime time, CalendarWeekRule rule, DayOfWeek firstDayOfWeek);
```

Unix `date` utility implements ISO 8601:
```
date +%V
```

In [PowerShell repo ](https://github.com/PowerShell/PowerShell/pull/6542) we use workaround described [here](https://blogs.msdn.microsoft.com/shawnste/2006/01/24/iso-8601-week-of-year-format-in-microsoft-net/) (see Keno's comment) for:
```powershell
Get-Date -UFormat %V
```

.Net should have parity and native implementation.

ISO 8601 is culture-independent so we can use static methods.

If we are going in the direction I believe it make sense consider all ISO8601 objects and compatibility with Unix `date` utility.
https://en.wikipedia.org/wiki/ISO_week_date#Relation_with_the_Gregorian_calendar
- first week
- last week
- weeks per year

http://man7.org/linux/man-pages/man1/date.1.html
- `%g` last two digits of year of ISO week number (see %G)
- `%G` year of ISO week number (see %V); normally useful only with %V
- `%V` ISO week number, with Monday as first day of week (01..53)

### Proposed API

```c#
namespace System.Globalization
{
    public abstract class Calendar : ICloneable
    {
        public static class ISOWeek
        {
            public static int GetWeekOfYear(DateTime time);
            public static int GetYear(DateTime time);
            public static DateTime GetYearStart(int year);
            public static DateTime GetYearEnd(int year);
            public static int GetWeeksInYear(int year);
            public static DateTime ToDateTime(int year, int week, DayOfWeek day);
    }
}
```

Please note that the proposed API is a static and not instance member. The reason is ISO week is culture and calendar independent.

### PR DESCRIPTION
**Pull request by khellang**
I'm not sure I've done everything correctly, but it seems to build and run locally. I guess we'll have to wait for https://github.com/dotnet/coreclr/pull/18456 to sync over before this'll build and run in CI.

Is it OK to place the tests in a single file, or should I create a folder and add separate files for each method, like some of the oter APIs?

Closes #28933

// @tarekgh

### CODE REVIEW

File: src/System.Globalization/ref/System.Globalization.Forwards.cs [hunk 1]
```diff
@@ -17,3 +17,4 @@
 [assembly: System.Runtime.CompilerServices.TypeForwardedTo(typeof(System.Globalization.RegionInfo))]
 [assembly: System.Runtime.CompilerServices.TypeForwardedTo(typeof(System.Globalization.StringInfo))]
 [assembly: System.Runtime.CompilerServices.TypeForwardedTo(typeof(System.Globalization.TextElementEnumerator))]
+[assembly: System.Runtime.CompilerServices.TypeForwardedTo(typeof(System.Globalization.ISOWeek))]
```
Comment (stephentoub) [inline src/System.Globalization/ref/System.Globalization.Forwards.cs]: Why is this needed?

Comment (stephentoub) [reply src/System.Globalization/ref/System.Globalization.Forwards.cs]: Related, which reference assembly is this being exposed from?  Presumably from System.Runtime, as that's where Calendar now is:
https://github.com/dotnet/corefx/blob/824af7f2ba689479767cddc33741cf030650d47e/src/System.Runtime/ref/System.Runtime.cs#L4096
so System.Runtime.cs should be updated to include the surface area for this new type.

File: src/System.Globalization/tests/ISOWeekTests.cs [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+using System.Linq;
+using Xunit;
+
+namespace System.Globalization
+{
+    public static class ISOWeekTests
+    {
+        // The following 71 years in a 400-year cycle have 53 weeks (371 days).
+        // Years not listed have 52 weeks (364 days). Add 2000 for current years.
+        private static readonly HashSet<int> s_YearsWith53Weeks = new HashSet<int>
```
Comment (stephentoub) [inline src/System.Globalization/tests/ISOWeekTests.cs]: Nit: Years => years

File: src/System.Globalization/tests/ISOWeekTests.cs [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+            224, 229, 235, 240, 246,
+            252, 257, 263, 268, 274,
+            280, 285, 291, 296, 303, 308, 314,
+            320, 325, 331, 336, 342,
+            348, 353, 359, 364, 370,
+            376, 381, 387, 392, 398
+        };
+
+        // From https://en.wikipedia.org/wiki/ISO_week_date#Relation_with_the_Gregorian_calendar.
+        private static readonly DateData[] s_DateData =
```
Comment (stephentoub) [inline src/System.Globalization/tests/ISOWeekTests.cs]: Nit: Date => date... will stop commenting on these... in general it should be s_camelCase.

File: src/System.Globalization/tests/ISOWeekTests.cs [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+            new YearData(2009, "2008-12-29", "2010-01-03"),
+        };
+
+        public static IEnumerable<object[]> WeeksInYear_TestData()
+        {
+            for (int year = 0; year <= 400; year++)
+            {
+                bool isLong = s_YearsWith53Weeks.Contains(year);
+                yield return new object[] { 2000 + year, isLong };
+            }
```
Comment (stephentoub) [inline src/System.Globalization/tests/ISOWeekTests.cs]: This one operation adds 400 unit tests that each show up individually, and it looks like the other theory inputs are similarly each adding a ton of inputs.  In the past we've had issues with too many tests bogging down the system, though that was when we were dealing with hundreds of thousands.  @Chrisboh, is that still a problem, such that we should consolidate all 400 of these inputs into one, or is it fine to have this?

File: src/System.Globalization/tests/ISOWeekTests.cs [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+        {
+            Assert.Equal(expected, ISOWeek.ToDateTime(year, week, dayOfWeek));
+        }
+
+        [Theory]
+        [InlineData(0)]
+        [InlineData(10000)]
+        public static void ToDateTime_WithInvalidYear_Throws(int year)
+        {
+            Assert.Throws<ArgumentOutOfRangeException>(() => ISOWeek.ToDateTime(year, 1, DayOfWeek.Friday));
```
Comment (stephentoub) [inline src/System.Globalization/tests/ISOWeekTests.cs]: For all of these argument exceptions, is there a parameter name that can be validated as well?  Assuming yes, it'd be good to do that and to switch to using AssertExtensions.Throws, which is like Assert.Throws in that it has an overload that accepts a parameter name, but it's able to not enforce that name on .NET implementations that may throw away metadata.

File: src/System.Globalization/ref/System.Globalization.Forwards.cs (hunk 1 above)
Comment (tarekgh) [reply src/System.Globalization/ref/System.Globalization.Forwards.cs]: yes we should expose this from System.Runtime 

Something like the following:

[https://github.com/dotnet/corefx/blob/master/src/System.Runtime/ref/System.Runtime.cs#L4096](https://github.com/dotnet/corefx/blob/master/src/System.Runtime/ref/System.Runtime.cs#L4096)

---
In reply to: [195475409](https://github.com/dotnet/corefx/pull/30386#discussion_r195475409) [](ancestors = 195475409)

Comment (tarekgh) [reply src/System.Globalization/ref/System.Globalization.Forwards.cs]: we should remove this from System.Globalization. having it in System.Runtime should be enough. and if we need to expose it later from any other contract, we can do it in System.Globalization.Calendars one. but I don't think we need to do it. 

cc @weshaggard

---
In reply to: [195474093](https://github.com/dotnet/corefx/pull/30386#discussion_r195474093) [](ancestors = 195474093)

**Comment by tarekgh** [comment]
> // Licensed to the .NET Foundation under one or more agreements.

could you please move this  test to System.Globalization.Calendars instead?

---
Refers to: src/System.Globalization/tests/ISOWeekTests.cs:1 in 2223d75. [](commit_id = 2223d751e68391bdc6db1b1208dfaaa7bfcd75cf, deletion_comment = False)

File: src/System.Globalization/tests/ISOWeekTests.cs [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+        {
+            Assert.Equal(expected, ISOWeek.GetYearEnd(year));
+        }
+
+        private static DateTime ParseIsoDate(string date)
+        {
+            return DateTime.ParseExact(date, "yyyy-MM-dd", CultureInfo.InvariantCulture);
+        }
+
+        private struct DateData
```
Comment (tarekgh) [inline src/System.Globalization/tests/ISOWeekTests.cs]: >DateData [](start = 23, length = 8)

why we need to use strings to parse it to initialize DateDate. can this be simplified by a constructor with the needed parameter? 


new DateDate(new DateTime(1, 1, 1), 1, 1, DayOfweek.Monday)? 

I know this is more typing but it is really simpler.

File: src/System.Globalization/tests/ISOWeekTests.cs (hunk 4 above)
Comment (tarekgh) [reply src/System.Globalization/tests/ISOWeekTests.cs]: The method using this data can be just converted to a Fact and internally can loop on the data.

---
In reply to: [195476775](https://github.com/dotnet/corefx/pull/30386#discussion_r195476775) [](ancestors = 195476775)

File: src/System.Globalization/tests/ISOWeekTests.cs (hunk 6 above)
Comment (khellang) [reply src/System.Globalization/tests/ISOWeekTests.cs]: We don't have to, I just thought it looked cleaner, but it should be quick to convert it 😊

**Comment by khellang** [comment]
> could you please move this test to System.Globalization.Calendars instead?

@tarekgh For some reason, I'm not able to pick up `ISOWeek` in `System.Globalization.Calendars.Tests`. Is there anything special I need to do?

**Comment by tarekgh** [comment]
@dotnet-bot test this please

**Comment by tarekgh** [comment]
@dotnet-bot test this please

**Comment by khellang** [comment]
Should I rebase this branch? Looks like there's some inconsistent CI results...

**Comment by tarekgh** [comment]
> Should I rebase this branch? Looks like there's some inconsistent CI results...

No, it looks there is something wrong with the CI runs.

@dotnet/dnceng could you please have a look?

**Comment by stephentoub** [comment]
@dotnet-bot test this please

**Comment by tarekgh** [comment]
```
D:\j\workspace\windows-TGrou---0d2c9ac4\Tools\ApiCompat.targets(58,5): error : TypesMustExist : Type 'System.Globalization.ISOWeek' does not exist in the implementation but it does exist in the contract. [D:\j\workspace\windows-TGrou---0d2c9ac4\src\System.Runtime\src\System.Runtime.csproj]
D:\j\workspace\windows-TGrou---0d2c9ac4\Tools\ApiCompat.targets(72,5): error : ApiCompat failed for 'D:\j\workspace\windows-TGrou---0d2c9ac4\bin\Windows_NT.AnyCPU.Debug\System.Runtime\uapaot\System.Runtime.dll' [D:\j\workspace\windows-TGrou---0d2c9ac4\src\System.Runtime\src\System.Runtime.csproj]
```

It looks we need to add exception in the apicompat filr for System.Runtime

**Comment by MattGal** [comment]
@tarekgh  taking a look...

**Comment by tarekgh** [comment]
@MattGal you may wait for a little as we asked for some other changes which will reset the CI. if we get the failures again, we can ping you again. Thanks.

**Comment by MattGal** [comment]
@tarekgh ; 2/3 of the failures are the apicompat issue you're already adressing.  The other looks like a random failure after the leaked dotnet.exe from the build causing Jenkins to not be able to clean up its workspace after build, not much we can do about that other than retry unless it is consistent.

File: src/System.Globalization.Calendars/tests/ISOWeek/ISOWeekTests.netcoreapp.cs [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+using System.Linq;
+using Xunit;
+
+namespace System.Globalization
+{
+    public static class ISOWeekTests
+    {
+        // The following 71 years in a 400-year cycle have 53 weeks (371 days).
+        // Years not listed have 52 weeks (364 days). Add 2000 for current years.
+        private static readonly HashSet<int> s_yearsWith53Weeks = new HashSet<int>
```
Comment (khellang) [inline src/System.Globalization.Calendars/tests/ISOWeek/ISOWeekTests.netcoreapp.cs]: Thinking I can just move this into a local now...

File: src/System.Globalization.Calendars/tests/ISOWeek/ISOWeekTests.netcoreapp.cs [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+            new DateData(new DateTime(2009, 12, 31), 2009, 53, DayOfWeek.Thursday),
+            new DateData(new DateTime(2010, 01, 01), 2009, 53, DayOfWeek.Friday),
+            new DateData(new DateTime(2010, 01, 02), 2009, 53, DayOfWeek.Saturday),
+            new DateData(new DateTime(2010, 01, 03), 2009, 53, DayOfWeek.Sunday),
+            new DateData(new DateTime(9999, 12, 31), 9999, 52, DayOfWeek.Friday),
+        };
+
+        private static readonly YearData[] s_yearData =
+        {
+            new YearData(2005, new DateTime(2005, 01, 03), new DateTime(2006, 01, 01)),
```
Comment (khellang) [inline src/System.Globalization.Calendars/tests/ISOWeek/ISOWeekTests.netcoreapp.cs]: I'll add cases for 0001 and 9999 as well...

File: src/System.Globalization.Calendars/tests/ISOWeek/ISOWeekTests.netcoreapp.cs (hunk 7 above)
Comment (tarekgh) [reply src/System.Globalization.Calendars/tests/ISOWeek/ISOWeekTests.netcoreapp.cs]: if you think we'll not add more test cases in the future which using it

---
In reply to: [196166639](https://github.com/dotnet/corefx/pull/30386#discussion_r196166639) [](ancestors = 196166639)

Comment (khellang) [reply src/System.Globalization.Calendars/tests/ISOWeek/ISOWeekTests.netcoreapp.cs]: Meh, I guess it doesn't matter where the data is. Might as just keep it as a field like this.

**Comment by khellang** [comment]
OK, I guess this is ready for another round 😄

**Comment by tarekgh** [comment]
@khellang I believe the thrown ArgumentOutOfRangeException is good here. you can just update your test to validate we are getting this exception. This is known limitation on DateTime and I am pretty sure having the end of the ISO year 9999 is not useful for anyone. but make sure you test 9998 though :-)

**Comment by khellang** [comment]
Anything left to do on this?

**Comment by tarekgh** [comment]
@khellang thanks getting this done. it is awesome!

**Comment by khellang** [comment]
@tarekgh Thank you! It was a pleasure 😁
